File: backend/app/schemas/drill.py
from datetime import datetime
from pydantic import Field, field_validator
from app.schemas.common import CamelModel


class DrillMetricsRequest(CamelModel):
    scan_id: str
    reaction_time_ms: int
    evacuation_time_ms: int
    posture_score_percentage: float
    completed_at_device: datetime

    @field_validator("reaction_time_ms")
    @classmethod
    def reaction_range(cls, value: int) -> int:
        if not 0 <= value <= 30 * 60 * 1000:
            raise ValueError("reactionTimeMs is out of range")
        return value

    @field_validator("evacuation_time_ms")
    @classmethod
    def evacuation_range(cls, value: int) -> int:
        if not 0 <= value <= 60 * 60 * 1000:
            raise ValueError("evacuationTimeMs is out of range")
        return value

    @field_validator("posture_score_percentage")
    @classmethod
    def posture_range(cls, value: float) -> float:
        if not 0 <= value <= 100:
            raise ValueError("postureScorePercentage must be between 0 and 100")
        return value


class DrillCompletionResponse(CamelModel):
    drill_id: str
    accepted: bool
    reward_eligible: bool
    safety_rating: float
    tier: str
    recorded_at: datetime

File: frontend/packages/contracts/src/schemas/drill.ts
import { z } from 'zod';

export const DrillSessionSchema = z.object({
  drillId: z.string(),
  scanId: z.string(),
  installationId: z.string(),
  startedAt: z.string().datetime(),
});
export type DrillSession = z.infer<typeof DrillSessionSchema>;

export const DrillMetricsSchema = z.object({
  scanId: z.string(),
  reactionTimeMs: z.number().int().min(0),
  evacuationTimeMs: z.number().int().min(0),
  postureScorePercentage: z.number().min(0).max(100),
  completedAtDevice: z.string().datetime(),
});
export type DrillMetrics = z.infer<typeof DrillMetricsSchema>;

export const DrillCompletionResponseSchema = z.object({
  drillId: z.string(),
  accepted: z.boolean(),
  rewardEligible: z.boolean(),
  safetyRating: z.number().min(0).max(100),
  tier: z.enum(['Platinum', 'Gold', 'Silver']),
  recordedAt: z.string().datetime(),
});
export type DrillCompletionResponse = z.infer<typeof DrillCompletionResponseSchema>;
